### PR TITLE
Implement Registering to the App

### PR DESCRIPTION
**Pull request by 3umarG**
## Summary

This PR implements the full cycle of user registeration, including all security configurations related to Spring Security with JWT.

## Changes checklist

- [X] Add Swagger dependency and test the localhost.
- [X] Build the first entity and table : `users`.
- [X] Add basic `AppUsersRepository` for dealing with CRUD opertations related to `AppUser` entity
.
- [X] Add jwt filter with security configurations.
- [X] Add `JwtService` with its implementation.
- [X] Add custom ApiResponse with some factories to generate the response.
- [X] Implement reigster endpoint.
- [X] Handle all edge cases for user registeration.

### CODE REVIEW

**Comment by 3umarG** [comment]
You can try register endpoint in swagger preview at : `http://localhost:9898/swagger-ui/index.html#/auth-controller/register`

**Comment by 3umarG** [comment]
The register dto should include : 
- **email** : unique + not null
- **userName** : unique + not null + used for login
- **password** : must be at least 8 characters long and contain at least one lowercase letter, one uppercase letter, one digit, and one special character (_regex_) .

File: build.gradle.kts [hunk 1]
```diff
@@ -36,13 +38,14 @@ dependencies {
     implementation("io.jsonwebtoken:jjwt-jackson:0.12.3")
 
     // Validation Dependencies
-    implementation("org.springframework.boot:spring-boot-starter-validation")
-    implementation("jakarta.validation:jakarta.validation-api:3.0.2")
+    implementation("org.springframework.boot:spring-boot-starter-validation:3.0.6")
+//    implementation("jakarta.validation:jakarta.validation-api:3.0.2")
```
Comment (xaaksw) [inline build.gradle.kts]: Can we remove this if not used ?

File: build.gradle.kts [hunk 2]
```diff
@@ -36,13 +38,14 @@ dependencies {
...
 
     // Swagger Dependencies
-    implementation("org.springdoc:springdoc-openapi-data-rest:1.6.0")
-    implementation("org.springdoc:springdoc-openapi-ui:1.6.0")
-    implementation("org.springdoc:springdoc-openapi-kotlin:1.6.0")
+//    implementation("org.springdoc:springdoc-openapi-data-rest:1.6.0")
+//    implementation("org.springdoc:springdoc-openapi-ui:1.6.0")
+//    implementation("org.springdoc:springdoc-openapi-kotlin:1.6.0")
```
Comment (xaaksw) [inline build.gradle.kts]: Same , try to avoid commenting code , if not used we can remove it

File: src/main/kotlin/com/re/back/dtos/request/auth/RegisterRequestDto.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import com.re.back.entities.auth.AppUser
+import jakarta.validation.constraints.*
+
+data class RegisterRequestDto(
+    @field:Email(
+        message = "Invalid email format. Please provide a valid email address."
+    )
+    @field:NotNull
+    @field:NotBlank
+    @field:NotEmpty
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/dtos/request/auth/RegisterRequestDto.kt]: We can only have notBlank validation here as it checks if the string is not null or empty

File: src/main/kotlin/com/re/back/dtos/request/auth/RegisterRequestDto.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    val userName: String,
+
+    @field:NotNull
+    @field:NotBlank
+    @field:NotEmpty
+    @field:Pattern(
+        regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@\$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$",
+        message = "Password must be at least 8 characters long and contain at least one lowercase letter, one uppercase letter, one digit, and one special character."
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/dtos/request/auth/RegisterRequestDto.kt:19]: This is not required in my pov , 
8 chars long is enough I guess

File: src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import java.time.LocalDateTime
+
+data class AuthResponseDto(
+    val id : Int?,
+    val userName : String,
+    val email : String,
+    val role : UserRole,
+    val bio : String?,
+    val token : String,
+    val expiresOn : LocalDateTime
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt]: Curious about why you return expires on?

File: src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.re.back.dtos.response
+
+import com.re.back.enums.UserRole
+import java.time.LocalDateTime
+
+data class AuthResponseDto(
+    val id : Int?,
+    val userName : String,
+    val email : String,
+    val role : UserRole,
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt]: Let's avoid roles for now, I guess now we only have user , and we can think about the roles in the future PR

File: src/main/kotlin/com/re/back/enums/UserRole.kt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package com.re.back.enums
+
+enum class UserRole {
+    ROLE_USER
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/enums/UserRole.kt:4]: What about USER?

File: src/main/kotlin/com/re/back/exceptions/CustomException.kt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+package com.re.back.exceptions
+
+abstract class CustomException(val statusCode: Int, message: String) : RuntimeException(message)
```
Comment (xaaksw) [inline src/main/kotlin/com/re/back/exceptions/CustomException.kt:3]: I recommend make the name of the exp more explained, maybe something like `AlreadyRegisterdException`

**Review by xaaksw — COMMENTED** [review]
Great work and really impressive effort , I think it's better to discuss this PR in a pairing session before merge, this is an add for me to understand auth! 
@3umarG 

Keep up the good work!

File: src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt (hunk 5 above)
Comment (3umarG) [reply src/main/kotlin/com/re/back/dtos/response/AuthResponseDto.kt]: That for frontend/mobile developers use to know when the token expired. Later we can add `/refresh-token` endpoint used by them.
We use this approach most of the time

File: src/main/kotlin/com/re/back/enums/UserRole.kt (hunk 7 above)
Comment (3umarG) [reply src/main/kotlin/com/re/back/enums/UserRole.kt:4]: That's  necessary in Spring Security context, I guess there is internal mechanism to determine the ROLE of the user if you defined specific role for specific endpoint, that allows this naming convention : `ROLE_{role_name}`

File: src/main/kotlin/com/re/back/exceptions/CustomException.kt (hunk 8 above)
Comment (3umarG) [reply src/main/kotlin/com/re/back/exceptions/CustomException.kt:3]: I make this a generic/abstract exception, and any specific exception will extend this exception type.
Actually I make this generic exp type to make it more standard in the context of exception handling.

**Example:**
- `AlreadyRegisteredException` & `NotMatchedPasswordException` are subtypes of `CustomException`; different in statusCode and message, but both have the same way of handling and same fields.

**Comment by 3umarG** [comment]
**Things we can discuss in the PR session :** 
- go more deep into Spring Security.
- talk about Response Factory VS Extension methods.

**Comment by 3umarG** [comment]
### Some edge cases handled in the scope of authentication and authorization :
#### 1. When user try to use _secured_ endpoint without token/authenticated.
- status code : **UnAuthorized 401**
- response body : 
```json
{
  "message": "You need to be Authorized to access this resource ...!!",
  "statusCode": 401,
  "isSuccess": false,
  "data": null
}
```

#### 2. Using expired jwt token :  
- status code : **Forbidden 403**
- response body : 
```json
{
  "message": "JWT Token is already expired ...!!",
  "statusCode": 403,
  "isSuccess": false,
  "data": null
}
```

#### 3. Not correct formatted/signature JWT token :
- status code : **Forbidden 403**
- response body : 
```json
{
  "message": "JWT Token has not correct format ...!!",
  "statusCode": 403,
  "isSuccess": false,
  "data": null
}
```

#### 4. The `sub` in the JWT token is deleted/not exist in the database
⚠️  This for the case if we delete/block user form the whole application and db, if he still has the generated token and it is still active, he can use it freely.

- status code : **UnAuthorized 401**
- response body : 
```json
{
  "message": "You are UN_AUTHORIZED of accessing this resource with exception : User Name not found",
  "statusCode": 401,
  "isSuccess": false,
  "data": null
}
```

#### Reason to go with all these edge cases ??
Spring Security -by default- doesn't give you this capabilities, it returns **Forbidden 403_** response if any of these cases happened, when dealing with mobile developers _specially_ they need to know more about api failure reasons. That will determine which path they will follow, specially in messages appear to end user, animation related to the failure reason … etc

**Comment by 3umarG** [comment]
## Integration Test for `/auth/register` endpoint

#### Test Cases :
- Including 3 test suits for most common 3 edge cases of the register endpoint
- **First** : `Register with valid inputs, returns 200 OK`
- **Second** : `Register with not valid input, returns 400 BAD_REQUEST`
- **Third** : `Register with taken username Or email, returns 400 BAD_REQUEST`

#### Approach & tools for testing : 
- Using basic _Assertions_ package for write assertions.
- Doesn't _Mock_ any service **yet !!**.
- Using `TestRestTemplate` with web environment `RANDOM_PORT` for calling endpoints.

**Review by xaaksw — APPROVED** [review]
a7la 7d yooom el 7d
